### PR TITLE
Parser test skeleton(Is this something we want?)

### PR DESCRIPTION
**Pull request by VCneverdie**
I wrote a draft of block parser test. Please take a look and let me know if we want something like this.

**I found assertion failure with this simple test. BlockParser seems cannot handle only TextBlock. Is that WAD?**

### CODE REVIEW

**Comment by andysim3d** [comment]
that's exactly what we want, good job 👍

**Comment by jinglongwu** [comment]
> 
> 
> I wrote a draft of block parser test. Please take a look and let me know if we want something like this.
> 
> **I found assertion failure with this simple test. BlockParser seems cannot handle only TextBlock. Is that WAD?**

It's not. It fails due to https://github.com/andysim3d/markdown/blob/master/block.py#L12.

When I design and implement the AST, I only updated `Element` class with its children structure but didn't notice `Block` has specific override for it. Please remove those overrides of `children()` and `nested()`

**Comment by VCneverdie** [comment]
> > I wrote a draft of block parser test. Please take a look and let me know if we want something like this.
> > **I found assertion failure with this simple test. BlockParser seems cannot handle only TextBlock. Is that WAD?**
> 
> It's not. It fails due to https://github.com/andysim3d/markdown/blob/master/block.py#L12.
> 
> When I design and implement the AST, I only updated `Element` class with its children structure but didn't notice `Block` has specific override for it. Please remove those overrides of `children()` and `nested()`

I think removing these two functions is not enough. Let's taking BlockParser as an example, if `begin = -1`, why we want to give -1 to `start_index` and `end_index`? Also the return is `content[end_index:]` in my opinion. Please correct me if I understand it wrong..

**Comment by jinglongwu** [comment]
not sure what you means above, but in our design, if the input of `BlockParser.parse()` is 'testing' as content, aka `BlockParser.parse('testing')`, the returned tree should be 
```
Element('testing')
      |
TextBlock('testing')
```

another example, if the input is as `BlockParser.parse('***abc***')`, which is a Bold and Italic, the returned tree should be
```
Element('***abc***')
      |
BoldBlock('*abc*')
      |
ItalicBlock('abc')
      |
TextBlock('abc')
```

Let me know whether this design makes sense to you.

with this design, the comparison in https://github.com/andysim3d/markdown/blob/441251722d822cdbfd3f6bc687483a1e5c9262c2/parser_test.py#L35

the `expected_root` should be  `Element('testing')`, and you also want to assert its children and children's contents.

**Comment by VCneverdie** [comment]
Re @jinglongwu 
Yeah I agree with this design and fix the assertion logic. However, I still think the logic `_invoke_parsers` in both ParagraphParser and BlockParser is not correct. 
For example: `BlockParser.parse('testing')`
```python
        for parser in self._parsers:
            begin, end, element = parser(content)
            # No matching parser for this content, begin can be -1
            if begin != 0 and begin < start_index:
                # inside this if, start_index, end_index, final_element = -1, -1, None
                start_index, end_index, final_element = begin, end, element 
        # at this time, start_index=-1
        if start_index != 0 and start_index != len(content):
            # error in content[0, -1]
            child = TextBlock(content[0, start_index])
```

File: parser_test.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+from img import ImgBlock
+from link_block import LinkBlock
+from codeblock import CodeBlock
+from fenched_code_block import FenchedCodeBlock
+from strikethrough_block import StrikethroughBlock
+
+from header import Header
+from horizontal_rule import HorizontalRule
+from list_paragraph import OrderedList, UnorderedList
+from quote import QuoteParagraph
```
Comment (VCneverdie) [inline parser_test.py:19]: The import will look better if we put elements in a package.

Comment (andysim3d) [reply parser_test.py:16]: > Re @jinglongwu
> Yeah I agree with this design and fix the assertion logic. However, I still think the logic `_invoke_parsers` in both ParagraphParser and BlockParser is not correct.
> For example: `BlockParser.parse('testing')`
> 
> ```python
>         for parser in self._parsers:
>             begin, end, element = parser(content)
>             # No matching parser for this content, begin can be -1
>             if begin != 0 and begin < start_index:
>                 # inside this if, start_index, end_index, final_element = -1, -1, None
>                 start_index, end_index, final_element = begin, end, element 
>         # at this time, start_index=-1
>         if start_index != 0 and start_index != len(content):
>             # error in content[0, -1]
>             child = TextBlock(content[0, start_index])
> ```

it fixed in #24

Comment (andysim3d) [reply parser_test.py:19]: add a issue #25

**Comment by VCneverdie** [comment]
Todo:
- add multi-level test cases

File: parser.py [hunk 2]
```diff
@@ -123,22 +125,30 @@ def parse(self, content, root=None):
 
         return root
 
-def create_paragrah_and_block_parsers():
```
Comment (VCneverdie) [inline parser.py:126]: I broke down this function so that `parser_test` can also reuse it.

File: parser.py [hunk 3]
```diff
@@ -20,11 +20,11 @@
 
 class AbstractParser(object):
     def __init__(self):
-        self._parsers = set()
+        self._parsers = dict()
 
     @property
     def parsers(self):
-        return self._parsers
+        return list(self._parsers)
```
Comment (VCneverdie) [inline parser.py]: Make parsers an ordered dict. It can keep the behavior like O(1) for register/unregister, no duplicate key and also have the order.

Comment (andysim3d) [reply parser.py]: 内啥，你可以直接用[ordereddict](https://docs.python.org/3/library/collections.html#collections.OrderedDict)

File: parser.py [hunk 4]
```diff
@@ -34,14 +34,14 @@ def register_parser(self, parser):
...
-        self._parsers.add(parser)
+        self._parsers[parser] = 0
 
     def unregister_parser(self, parser):
         '''
         Remove a registered parser
         '''
-        if parser in self._parsers:
-            self._parsers.remove(parser)
+        if parser in self._parsers.keys():
```
Comment (andysim3d) [inline parser.py]: parser in self._parsers

**Review by andysim3d — APPROVED** [review]

**Pull request merged**